docs/manual/technical_design.md

```diff
@@ -0,0 +1,334 @@
+
+This action requires a lock.
+
+### BounceProcesses
+
+The `BounceProcesses` subreconciler restarts any `fdbserver` processes that do not have the correct command line. This is done through the `kill` command in fdbcli, which causes the processes to immediately exit, which causes `fdbmonitor` to restart them. This will restart any process for a process group that has the `IncorrectCommandLine` condition.
+
+When upgrading a cluster to a new version of FoundationDB, we follow a special process. In most cases, each instance of the operator only restarts processes that are under its control, which means that in multi-KC clusters we will restart processes in multiple batches, with one batch for each KC. During an upgrade, we cannot use this strategy, because protocol-incompatible upgrades require all processes to be updated simultaneously. To make this work, we have each instance of the operator use the locking system to store a list of processes that it has prepared for the upgrade in the database. Each instance of the operator then checks that list and compares it against the database status to confirm that every process that is reporting to the database is ready for the upgrade. It will then restart all of the processes across the entire cluster and move forward with its own reconciliation. When the other instances of the operator run their next reconciliation, they will see that the processes they are managing have the correct command-line, and will move past the bounce stage.
+
+If a process needs to be restarted but is not reporting to the database, this will requeue reconciliation with an error.
+
+This will not attempt to restart any process that is flagged for removal.
+
+This will not restart processes until every process has been up for 600 seconds. This limit can be configured through the `minimumUptimeSecondsForBounce` field in the cluster spec.
+
+This action requires a lock.
+
+### UpdatePods
+
+The `UpdatePods` subreconciler deletes any pods that have incorrect pod specs. Once it deletes a pod, it will requeue reconciliation so that the operator can recreate the pod on the next reconciliation run.
+
+This will only delete pods with a single `zoneid` locality value, which ensures that we only lose one unit of fault tolerance through these restarts.
+
+This will not delete any pods that are flagged for removal.
+
+If any pod is in a terminating state and is not flagged for removal, this will not delete any further pods. It will requeue reconciliation until the in-flight termination completes.
+
+This action requires a lock.
+
+### RemoveServices
+
+The `RemoveServices` subreconciler deletes any services that are no longer required for the cluster.
+
+### RemovePods
+
+The `RemovePods` subreconciler deletes any pods that are marked for removal and have been fully excluded, meaning that they are not serving any roles or holding any data.
```

Comment:
I'll open an issue to rename this to `RemoveProcessGroup` (https://github.com/FoundationDB/fdb-kubernetes-operator/issues/790).